src/shared/discovery.js

```diff
@@ -104,20 +104,34 @@ export default class Discovery {
     let beaconMessage;
     if (this.server) {
       beaconMessage = '__cross_frame_dhcp_offer';
+
+      // The `sidebar` iframe is the server and its parent is the `host` frame.
+      this.target.parent.postMessage(beaconMessage, this.origin);
     } else {
       beaconMessage = '__cross_frame_dhcp_discovery';
-    }
 
-    // Perform a top-down, breadth-first traversal of frames in the current
-    // window and send messages to them.
-    const queue = [this.target.top];
-    while (queue.length > 0) {
-      const parent = /** @type {Window} */ (queue.shift());
-      if (parent !== this.target) {
-        parent.postMessage(beaconMessage, this.origin);
-      }
-      for (let i = 0; i < parent.frames.length; i++) {
-        queue.push(parent.frames[i]);
+      // Three possibilities here: the `notebook` iframe, an annotatable iframe(s), or the `host` frame
+      if (
+        // Both the `notebook` iframe and annotatable iframe(s) have as parent the `host` frame.
+        //  - the `notebook` iframe is identified by having an `hypothesis-app` element
+        //  - annotatable iframe(s) are identified by having a global `window.__hypothesis_frame` variable
+        this.target.document.querySelector('hypothesis-app') ||
```

Comment:
In my testing this code path is never invoked for the notebook app because the notebook never invokes the discovery process by calling `Discovery#connect`.